nixos/modules/services/networking/moodle-dl.nix

```diff
@@ -0,0 +1,63 @@
+{ config, lib, pkgs, ... }:
+with lib;
+let
+  cfg = config.services.moodle-dl;
+  moodle-dl-json = pkgs.writeText "moodle-dl.json" (builtins.toJSON cfg.config);
+in {
+  options = {
+    services.moodle-dl = {
+      enable = mkEnableOption "moodle-dl, a Moodle downloader";
+
+      config = mkOption {
+        default = {};
+        type = types.attrs;
+        description = ''
+          Configuration for moodle-dl. For a full example, see
+          <link xlink:href="https://github.com/C0D3D3V/Moodle-Downloader-2/wiki/Config.json"/>.
+        '';
+      };
+
+      startAt = mkOption {
+        type = with types; either str (listOf str);
+        description = "When to run moodle-dl. See systemd.time(7) for the format.";
+      };
+
+      directory = mkOption {
+        default = "/var/lib/moodle";
+        type = types.path;
+        description = "The path moodle-dl should download course files to.";
+      };
+
+      package = mkOption {
+        default = pkgs.moodle-dl;
+        type = types.package;
+        description = "The moodle-dl package to use.";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+
+    users.extraUsers.moodle = {
+      isNormalUser = false;
+      home = cfg.directory;
+      createHome = true;
```

Comment:
I'm not fully clear on what the code does. It seems it creates a `StateDirectory` for `moodle` (this will be `/var/lib/moodle`?) if `cfg.directory` is set?

Reply:
Pardon my silly typo! I missed a pretty critical part! I have updated the example. The snippet of code will create `/var/lib/moodle-dl` **if** (and only if) the sysadmin leaves `directory` as the default value. Additional to this you could adjust the `description` of the `directory` option to something like this:

```
The directory moodle-dl should download course files to. If left
as the default value this directory will automatically be created before
moodle-dl runs, otherwise the sysadmin is responsible for ensuring
the directory exists with appropriate ownership and permissions.
```

Reply:
Thank you for clearing this up! I'll follow your suggestion, using `StateDirectory = "moodle-dl"` to avoid conflics with the moodle module.

But does systemd guarantee that the StateDirectory will exist before ExecStartPre is run? (Which copies the configuration file to `cfg.directory`)

Reply:
Yes.